完善长期记忆重复性检测

Hoshino-Yumetsuki opened the pull request:
- 添加使用三种字符串相似度检测算法以准确的检测记忆是否重复
    * 余弦相似度：考虑词频信息
    * Levenshtein距离：考虑字符级别的编辑操作
    * Jaccard相似度：考虑词集合的重叠程度

    在实际测试中比单纯的向量相似度检测更为准确且好用

- 修复 Milvus 缺少记忆 ID 问题

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
	- Introduced a new long-term memory plugin with configurable similarity checks.
	- Added advanced text similarity calculation functionality.
	- Enhanced memory management with configurable thresholds.

- **Configuration Changes**
	- Removed legacy long-term memory configuration options.
	- Added new `enableSimilarityCheck` configuration option.
	- Simplified configuration schema for memory management.

- **Documentation**
	- Updated README for the long-memory plugin.
	- Added badges and description for the new plugin.

- **Refactor**
	- Streamlined memory handling across multiple components.
	- Removed deprecated long-memory related code.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Hoshino-Yumetsuki committed:
feat(core): improve memory similarity filtering with cosine similarity calculation

- Refactor `filterSimilarMemory` function to use a cosine similarity threshold for filtering memories.
- Increase the number of similar memories retrieved from the vector store from 10 to 20.
- Introduce a new `cosineSimilarity` function to calculate the similarity score between memory strings.
- Update logging to warn when a memory is too similar based on the cosine score.

Hoshino-Yumetsuki committed:
fix(core): adjust similarity threshold in memory filtering

- Decrease the number of similar memories retrieved from the vector store from 20 to 10 in the `filterSimilarMemory` function.
- This change aims to refine the memory filtering process for improved performance and relevance.

Hoshino-Yumetsuki committed:
refactor(core): replace cosine similarity function with calculateSimilarity utility

- Updated the `filterSimilarMemory` function to utilize the new `calculateSimilarity` utility for improved memory similarity scoring.
- Enhanced logging to provide detailed similarity results, including score and additional details.
- Removed the previous `cosineSimilarity` function to streamline the codebase.

Hoshino-Yumetsuki committed:
feat(core): add similarity check configuration and refine memory filtering logic

- Introduced `enableSimilarityCheck` option in the configuration to control memory similarity checks.
- Updated the `filterSimilarMemory` function to utilize the new configuration, allowing for more flexible memory filtering based on similarity thresholds.
- Adjusted the condition for filtering similar memories to improve performance and relevance.

Hoshino-Yumetsuki committed:
feat(core): update similarity check configuration and enhance documentation

- Added `enableSimilarityCheck` option to the configuration schema to control memory similarity checks.
- Updated English and Chinese locale files to include descriptions for the new `enableSimilarityCheck` and refined `longMemoryAddSimilarity` settings.
- Improved clarity in documentation regarding the behavior of memory addition based on similarity thresholds.

Hoshino-Yumetsuki committed:
refactor(core): streamline configuration and improve memory filtering logic

- Simplified the configuration schema by consolidating the `enableSimilarityCheck` option.
- Enhanced the `filterSimilarMemory` function by removing unnecessary line breaks for better readability.
- Improved logging format for similarity checks to ensure clarity in output.

Hoshino-Yumetsuki committed:
feat(vector-store): add langchain_primaryid to document metadata in apply function

- Enhanced the document metadata structure by adding a new field `langchain_primaryid` alongside the existing `raw_id`.
- This change improves the identification and tracking of documents within the vector store service.

dingyi222666 commented on packages/core/src/config.ts:
请以 longMemory 开头

dingyi222666 commented on packages/core/src/llm-core/memory/history/similarity.ts:
是否考虑多语言支持?中日英分词效果呢?

dingyi222666 commented on packages/vector-store-service/src/vectorstore/milvus.ts:
参考 https://github.com/langchain-ai/langchainjs/blob/45498632ce2f5d539d84d049bf5b6717f674ac46/libs/langchain-community/src/vectorstores/milvus.ts#L268，不应添加此项，传递 ids 后上游代码已经处理

dingyi222666 commented on packages/core/src/llm-core/memory/history/similarity.ts:
```diff
@@ -0,0 +1,108 @@
+export interface SimilarityResult {
```
不要放在此处，应该放在 https://github.com/ChatLunaLab/chatluna/tree/v1-dev/packages/core/src/llm-core/utils

Hoshino-Yumetsuki committed:
refactor(vector-store): remove langchain_primaryid from document metadata in apply function

- Eliminated the `langchain_primaryid` field from the document metadata structure in the `apply` function.
- This change simplifies the metadata and focuses on the essential `raw_id` for document identification.

Hoshino-Yumetsuki committed:
refactor(core): update long memory configuration and filtering logic

- Renamed `longMemoryAddSimilarity` to `longMemoryDuplicateThreshold` for clarity in configuration.
- Introduced `longMemoryEnableDuplicateCheck` boolean option to control duplicate memory checks.
- Updated the `apply` function to utilize the new configuration options for filtering similar memories.
- Improved readability by renaming parameters in the `clear-chat-history` event handler.

Hoshino-Yumetsuki committed:
feat(core): add new dependencies and refactor similarity calculation

- Added new dependencies: `@node-rs/jieba` and `tiny-segmenter` to enhance text processing capabilities.
- Refactored the import path for the `calculateSimilarity` function to improve code organization.
- Removed the obsolete `similarity.ts` file, consolidating similarity calculation logic for better maintainability.

Hoshino-Yumetsuki committed:
feat(core): enhance similarity calculation and add stopwords filtering

- Introduced `stopwords-iso` dependency to improve text processing by filtering out common stopwords in multiple languages.
- Updated similarity weights in the `SimilarityCalculator` to include a new `bm25` metric, adjusting existing weights for cosine, levenshtein, and jaccard.
- Refactored `TextTokenizer` to detect languages and remove stopwords from tokenized text, enhancing the accuracy of similarity calculations.
- Added `calculateBM25Similarity` method to compute similarity scores using the BM25 algorithm, providing a more robust similarity assessment.

Hoshino-Yumetsuki committed:
refactor(core): standardize method calls in TextTokenizer and SimilarityCalculator

- Updated method calls in the TextTokenizer class to use static references for language detection and stopword filtering.
- Refactored SimilarityCalculator to call its similarity methods statically, improving code clarity and consistency.
- These changes enhance maintainability and readability of the similarity calculation logic.

Hoshino-Yumetsuki committed:
feat(core): validate and adjust similarity weights for improved accuracy

- Introduced a new function `validateAndAdjustWeights` to ensure that the total similarity weights sum to 1, adjusting them if necessary.
- Updated the `SimilarityCalculator` to use the validated weights in the similarity score calculation, enhancing the reliability of similarity assessments.
- This change improves the robustness of the similarity calculation logic by ensuring consistent weight application.

dingyi222666 committed:
feat(long-memory): add long-term memory plugin and configuration support

dingyi222666 committed:
refactor(long-memory): update plugin imports and standardize type definitions

dingyi222666 committed:
refactor(long-memory): rename duplicate check configuration for clarity

dingyi222666 committed:
refactor(long-memory): remove longMemory flag from configuration for simplification

dingyi222666 committed:
refactor(long-memory): remove unnecessary comment from index.ts for clarity
